src/Microsoft.Azure.SignalR.Common/ClientInvocation/ClientResultsManager.cs

```diff
@@ -0,0 +1,312 @@
+        private ulong _lastInvocationId = new();
+
+        private readonly ConcurrentDictionary<string, (Type Type, string ConnectionId, string CallerServerId, object Tcs, Action<object, CompletionMessage> Complete)> _routedInvocations = new();
+
+        private readonly ConcurrentDictionary<string, ServiceMappingMessage> _serviceMappingMessages = new();
+        private readonly IHubProtocolResolver _hubProtocolResolver;
+
+        public ClientResultsManager(IHubProtocolResolver hubProtocolResolver)
+        {
+            _hubProtocolResolver = hubProtocolResolver;
+        }
+
+        public ulong GetNewInvocation()
+        {
+            return Interlocked.Increment(ref _lastInvocationId);
+        }
+
+        public void AddServiceMappingMessage(string invocationId, ServiceMappingMessage serviceMappingMessage)
+        {
+            _serviceMappingMessages.TryAdd(invocationId, serviceMappingMessage);
+        }
+
+        public void RemoveServiceMappingMessageWithOfflinePing(string instanceId)
+        {
+            foreach (var (_, serviceMappingMessage) in _serviceMappingMessages)
+            {
+                if (serviceMappingMessage.InstanceId == instanceId)
+                {
+                    if (_pendingInvocations.TryRemove(serviceMappingMessage.InvocationId, out var item))
+                    {
+                        var message = new CompletionMessage(serviceMappingMessage.InvocationId, $"Connection '{serviceMappingMessage.ConnectionId}' disconnected.", null, true);
+                        item.Complete(item.Tcs, message);
+                    }
+                }
+            }
+        }
+
+        public Task<object> AddRoutedInvocation(string connectionId, string invocationId, string callerServerId, CancellationToken cancellationToken)
+        {
+            var tcs = new TaskCompletionSourceWithCancellation<object>(this, connectionId, invocationId, cancellationToken);
+            var result = _routedInvocations.TryAdd(invocationId, (typeof(object), connectionId, callerServerId, tcs, static (state, completionMessage) =>
+            {
+                var tcs = (TaskCompletionSourceWithCancellation<object>)state;
+                if (completionMessage.HasResult)
+                {
+                    tcs.SetResult(completionMessage.Result);
+                }
+                else
+                {
+                    tcs.SetException(new Exception(completionMessage.Error));
+                }
+            }
+            ));
+            Debug.Assert(result);
+
+            tcs.RegisterCancellation();
+
+            return tcs.Task;
+        }
+
+        public Task<T> AddInvocation<T>(string connectionId, string invocationId, CancellationToken cancellationToken)
+        {
+            var tcs = new TaskCompletionSourceWithCancellation<T>(this, connectionId, invocationId, cancellationToken);
```

Comment:
is canellationToken.Register(()=>tcs.SetCancelled()) enough?

Reply:
See [this comment](https://github.com/Azure/azure-signalr/pull/1684/files/7a1316985004302c680a5519cec9ea85d97950d9#r979471825)